# Discrete-Time Clock Hands

Total seconds in 12 hours: $60 \cdot 60 \cdot 12 = 43200$

Define a unit _tick_ during which one second passes.  There are $60 \cdot 60 \cdot 12 = 43200$ ticks in 12 hours; so that is the number of distinct states of the clock face.

Define three variables $s$, $m$, and $h$, which represent the position of the second, minute, and hour hand, respectively, measured in ticks, at time $t$ seconds.  Then:

The hour hand makes one revolution of the clock face every 12 hours, so each second it moves a single tick:
$h = t \bmod 43200$.

The minute hand makes 12 revolutions of the clock face every 12 hours:
$m = 12 t \bmod 43200$.

Finally, the second hand whooshes around the clock face 60 times a minute:
$s = 60 m = 720 t \bmod 43200$.

So, for a few examples,
 - at 12:00:00, $h = m = s = 0$.
 - at 12:00:01, $h = 1$, $m = 12$, and $s = 720$.
 - at 12:01:00, $h = 60$, $m = 720$, and $s = 0$.
 - at 1:05:05, $t = 3600 + 300 + 5 = 3905$, so $h = 3905$, $m = 3660$, and $s = 3600$.

That should be enough to code for the solution.

## Code!

In [37]:
TICKS = 43200

class Hands:
    """Compute angles of clock hands in units of ticks (2pi / 60 / 60 / 12)."""
    def __init__(self, seconds):
        """Time in seconds uniquely determines angle of each clock hand."""
        self.h = seconds % TICKS
        self.m = (12 * seconds) % TICKS
        self.s = (720 * seconds) % TICKS

        # This computation of distance between hands is not right; it does not deal 
        # with cases where the correct difference spans the origin.  Fortunately,
        # this does not affect the problem at hand (I hope).
        self.d = int(max(self.h, self.m, self.s)
                     - min(self.h, self.m, self.s))
        
    def __str__(self):
        """Dump state."""
        return '{0:d} {1:d} {2:d} {3:d}'.format(self.h, self.m, self.s, self.d)
    
    def clock(self):
        """Convert to HH:MM:SS format."""
        return '{0:02d}:{1:02d}:{2:02d}'.format(self.h // 3600, (self.h // 60) % 60, self.h % 60)

# Find hand angles (hms values) for each clock state.
hands = [Hands(t) for t in range(TICKS)]

# Display the minimum value, excluding 12:00:00 which we already know to be zero.
print('Minimum angular separation in ticks: {0:d}\n'.format(min([tt.d for tt in hands[1:]])))

# Display a few of the smallest values.
MINDELTA = 300
print('Hand states with separation of at most {0:d} ticks:'.format(MINDELTA))
for tt in hands:
    if MINDELTA >= tt.d:
        print('\t{0:s} is {1:s}'.format(tt.clock(), str(tt)))

Minimum angular separation in ticks: 181

Hand states with separation of at most 300 ticks:
	00:00:00 is 0 0 0 0
	02:11:11 is 7871 8052 7920 181
	03:16:16 is 11776 11712 11520 256
	08:43:44 is 31424 31488 31680 256
	09:48:49 is 35329 35148 35280 181


## Voila!

The minimum angular separation of clock hands, of zero degrees, occurs at high noon.  But we knew that.

The minimum nonzero angular separation happens at **2:11:11** and **9:48:49**, of $181 \cdot 2\pi / 43200 \approx 0.026$ radians, or a skosh more than $1.5$ degrees.

Runner-up is at **3:16:16** and **8:43:44** with the hands $2.133\ldots$ degrees apart.

## Footnote

Yes, I cheated and wrote a program.  An analytic solution to the minimization problem would be *much* harder.  I think the approach would use a difference equation?  It would be fun to Go There some day, but not this day.